packages/hardhat-core/src/types/plugins.ts

```diff
@@ -0,0 +1,15 @@
+import { DeepReadonly } from "ts-essentials";
```

Comment:
I think this module shouldn't be named `plugins`, but probably `config-dsl`. 

The only place that makes these things public is `hardhat-core/src/internal/core/config/config-env.ts`.

Reply:
What about just moving it to config then?

I think the term DSL just adds unnecessary complexity, especially when it's something that lives in a gray zone between a "proper" DSL an an API.

Reply:
Moving it to the `config` module introduces a cyclic dependency with the `runtime` module. Maybe it could be moved to the `runtime` module instead. But I'm not sure that's for the best.

Reply:
Oh, you are right, @fvictorio. We decided to stop using the word DSL some time ago. I forgot about it.

Maybe `config-env`?

Reply:
The functions that use these types are currently in `config-env` so it might work. I'm not sure it's a good name but I can't think of another one.

Reply:
Ok, what about removing this module, moving `ConfigExtender` to `config` and `EnvironmentExtender` to `runtime`? As far as I can tell, that doesn't introduce any cycles.

Reply:
lgtm